### DISCUSSION TITLE
Will this work on G3 3590 from Dell?

### DISCUSSION THREAD

**Discussion by tamimmahmudkhan** [DadouLapous/DellG5SE-Fan-Linux#4] [Q&A] [answered]
Will this work on G3 3590 from Dell?

**Comment by DavidLapous** [comment, marked as answer]
From the pictures I'm seeing, this laptop uses a very similar cooling solution to the Dell G5 SE's. There is a good chance it will work perfectly. If you can provide me an EC dump with either : [RWEverything](https://www.majorgeeks.com/files/details/rweverything.html) on windows (search for the EC icon), or, with linux,  [isw](https://github.com/YoyPa/isw) (with `isw -c`), or [nbfc](https://github.com/hirschmann/nbfc) (with `ec-probe dump`) or compile [this linux kernel function](https://github.com/torvalds/linux/blob/master/tools/power/acpi/tools/ec/ec_access.c)  (and get a dump with `./ec_access -r`), I should be able to confirm that your [EC](https://en.wikipedia.org/wiki/Embedded_controller) uses the same offset (ie. fan variable addresses) as my laptop and you'll safely be able to use this tool.

**Reply by tamimmahmudkhan** [reply]
This is what isw dumped 

`EC dump 
       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
od: /sys/kernel/debug/ec/ec0/io: No such file or directory`

**Reply by DavidLapous** [reply]
The module `ec_sys` is not loaded by your kernel. Does `grep ec_sys /lib/modules/$(uname -r)/modules.builtin` outputs something ?

**Reply by tamimmahmudkhan** [reply]
Nope. I am running base Arch atm with i8kutils for fan control.

**Reply by DavidLapous** [reply]
Okay, can you try : `sudo modprobe ec_sys write_support=1 `, and then retry  dumping with isw ?

**Reply by tamimmahmudkhan** [reply]
```EC dump 
       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
000000 33 00 04 00 00 01 03 00 00 00 00 00 00 00 00 00  >3...............<
000010 00 64 00 00 73 31 7a 0d 15 0c 88 47 08 00 7a 0d  >.d..s1z....G..z.<
000020 7a 11 88 2c 03 4f 11 00 0b 08 00 00 00 00 d1 ff  >z..,.O..........<
000030 00 01 01 03 36 ff ff 02 00 00 00 00 00 00 21 00  >....6.........!.<
000040 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  >................<
000050 00 00 00 00 00 00 00 b0 12 cc 09 00 00 00 00 00  >................<
000060 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  >................<
*
000100```

**Reply by DavidLapous** [reply]
I don't know why isw doesn't give the full dump (lines 7 to F are not shown, * instead...) Can you try with `nbfc` (also in the AUR)?

**Reply by tamimmahmudkhan** [reply]
```
  | 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
---|------------------------------------------------
00 | 33 00 04 00 00 01 03 00 00 00 00 00 00 00 00 00 
10 | 00 64 00 00 6B 31 7A 0D 12 0C 88 47 08 00 7A 0D 
20 | 7A 11 88 2C 03 4F 11 00 0B 08 00 00 00 00 D1 FF 
30 | 00 01 01 03 36 FF FF 02 00 00 00 00 00 00 21 00 
40 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
50 | 00 00 00 00 00 00 00 92 12 EE 12 00 00 00 00 00 
60 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
70 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
80 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
90 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
A0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
B0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
C0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
D0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
E0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
F0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
```

**Reply by DavidLapous** [reply]
Thanks, so that was only zeros ^^. So your laptop doesn't have the same offsets as mine at all. Fan management seems approximately the same, but I don't know where your manual mode is... Sadly,  you'll have to try reverse engineering that by yourself since I don't own this laptop. 
How to do that : use `ec-probe monitor` to find which variables are changing over time. 
Try to first find the offsets holding the current cpu / gpu fan speed (by changing fan speed with i8kutils for instance, or stressing cpu/gpu, or doing it with windows if dell let you control your fans). Next to these offsets, there should be a variable that hold the target fan speed.  *Usually*, fan turned off is `0xFF` (for both target and current fan speed).
Once you've find them, try to write values to theses offset (preferably seen values) to confirm. 
That may not be working as a manual mode may be needed (that is another offset to find). 
If you think you have made a mistake somewhere, reboot. EC variables gets reset upon reboot.

**Reply by tamimmahmudkhan** [reply]
Okay thanks for your time. I will try it if I can't find a better solution.
